### PR TITLE
EAP7-1594 Integrate MP Fault Tolerance 3.0 into EAP

### PR DESCRIPTION
**Pull request by rhusar**
Replaces
https://github.com/wildfly/wildfly-proposals/pull/361

Jira
https://issues.redhat.com/browse/EAP7-1594

### CODE REVIEW

File: microprofile/WFLY-14281_Integrate-MP-Fault-Tolerance-3.0.adoc [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+MicroProfile 4.0 provides a major upgrade for https://github.com/eclipse/microprofile-fault-tolerance/releases/tag/3.0[MicroProfile Fault Tolerance 3.0] with changes described in its
+https://download.eclipse.org/microprofile/microprofile-fault-tolerance-3.0/microprofile-fault-tolerance-spec-3.0.html#release_notes_30[Release Notes].
+
+This component upgrade also covers the upgrade of SmallRye Fault Tolerance that implements MicroProfile Fault Tolerance 3.0.
+
```
Comment (fabiobrz) [inline microprofile/WFLY-14281_Integrate-MP-Fault-Tolerance-3.0.adoc:15]: I think that here in the overview it should be noted here the fact that the major upgrade will introduce breaking changes in the spec itself and also in the integration with the Metrics implementation which is provided in the same runtime context (i.e. WildFly). 
What do you think about mentioning it here and move the "Breaking changes" section here for it to be a subsection of the "Overview"?

File: microprofile/WFLY-14281_Integrate-MP-Fault-Tolerance-3.0.adoc [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+== Issue Metadata
+ 
+=== Issue
+
+* https://issues.jboss.org/browse/WFLY-14281[WFLY-14281 Integrate MP Fault Tolerance 3.0]
+
+=== Related Issues
+
+* https://issues.redhat.com/browse/EAP7-1594[EAP7-1594 Integrate MP Fault Tolerance 3.0 into EAP]
+
```
Comment (fabiobrz) [inline microprofile/WFLY-14281_Integrate-MP-Fault-Tolerance-3.0.adoc:25]: As properly stated in the "Braking changes" section at the end of the proposal, shouldn't relevant Metrics related issues be mentioned here? 
BTW - Is the issue for Metrics update blocking this one? In such case mentioning this fact is also needed.
WDYT?

File: microprofile/WFLY-14281_Integrate-MP-Fault-Tolerance-3.0.adoc [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+* mailto:fburzigo@redhat.com[Fabio Burzigotti]
+
+=== Testing By
+
+* [x] Engineering
+
+* [ ] QE
+
+=== Affected Projects or Components
+
```
Comment (fabiobrz) [inline microprofile/WFLY-14281_Integrate-MP-Fault-Tolerance-3.0.adoc:42]: Can links be added to the list items here and below? It helps the reader browse to the exact content the writer has in mind when providing a reference..

File: microprofile/WFLY-14281_Integrate-MP-Fault-Tolerance-3.0.adoc [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+* MicroProfile
+
+== Requirements
+
+* Upgrade MicroProfile Fault Tolerance (specification) to 3.0
+* Upgrade smallrye-fault-tolerance to version 5.x (which implements the above specification)
+* Run all MicroProfile Fault Tolerance tests in WildFly integration test suite
+* Run the Eclipse MicroProfile Fault Tolerance TCK without failures
+
```
Comment (fabiobrz) [inline microprofile/WFLY-14281_Integrate-MP-Fault-Tolerance-3.0.adoc:55]: There is no need to add tests regarding the integration with the upgraded Metrics subsystem? E.g.: shouldn't there be tests that ensure compatibility between the two subsystem, like for instance 
```
if (MP-FT >= 3) then assert(MP-Metrics >= 2)
```
?

File: microprofile/WFLY-14281_Integrate-MP-Fault-Tolerance-3.0.adoc [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+=== Nice-to-Have Requirements
+
+=== Non-Requirements
+
+== Test Plan
+
+* Run existing tests (part of the CI runs):
+** `wildfly/testsuite/integration/microprofile`
+** `wildfly/testsuite/integration/microprofile-tck/fault-tolerance`
+
```
Comment (fabiobrz) [inline microprofile/WFLY-14281_Integrate-MP-Fault-Tolerance-3.0.adoc:70]: See my comments above regarding testing the integration between Metrics and FT.

**Review by fabiobrz — CHANGES_REQUESTED** [review]
Thanks for this MR @rhusar - I've dropped some comments and questions. Let me know what you think.

File: microprofile/WFLY-14281_Integrate-MP-Fault-Tolerance-3.0.adoc (hunk 1 above)
Comment (rhusar) [reply microprofile/WFLY-14281_Integrate-MP-Fault-Tolerance-3.0.adoc:15]: Yeah, moved from bottom to here.

File: microprofile/WFLY-14281_Integrate-MP-Fault-Tolerance-3.0.adoc (hunk 2 above)
Comment (rhusar) [reply microprofile/WFLY-14281_Integrate-MP-Fault-Tolerance-3.0.adoc:32]: Nothing in metrics is blocking us, it's just they needed to go in 'together'.

File: microprofile/WFLY-14281_Integrate-MP-Fault-Tolerance-3.0.adoc (hunk 3 above)
Comment (rhusar) [reply microprofile/WFLY-14281_Integrate-MP-Fault-Tolerance-3.0.adoc:49]: Link to what?

File: microprofile/WFLY-14281_Integrate-MP-Fault-Tolerance-3.0.adoc (hunk 4 above)
Comment (rhusar) [reply microprofile/WFLY-14281_Integrate-MP-Fault-Tolerance-3.0.adoc:62]: This is covered by the existing tests and the TCK. It's not like there is a support metrics, its 'all or nothing' as in 'MP4 platform'.

File: microprofile/WFLY-14281_Integrate-MP-Fault-Tolerance-3.0.adoc (hunk 5 above)
Comment (rhusar) [reply microprofile/WFLY-14281_Integrate-MP-Fault-Tolerance-3.0.adoc:77]: Those are covered.

File: microprofile/WFLY-14281_Integrate-MP-Fault-Tolerance-3.0.adoc (hunk 3 above)
Comment (fabiobrz) [reply microprofile/WFLY-14281_Integrate-MP-Fault-Tolerance-3.0.adoc:49]: Are the Quickstarts affected in any way by the upgrade?

File: microprofile/WFLY-14281_Integrate-MP-Fault-Tolerance-3.0.adoc (hunk 2 above)
Comment (fabiobrz) [reply microprofile/WFLY-14281_Integrate-MP-Fault-Tolerance-3.0.adoc:32]: From https://download.eclipse.org/microprofile/microprofile-fault-tolerance-3.0/microprofile-fault-tolerance-spec-3.0.html#release_notes_30:

```
The metrics added automatically by MicroProfile Fault Tolerance have been updated to take advantage of support for metric tags which was added to MicroProfile Metrics in version 2.0
```

So FT 3.0 uses Metrics 2.0 APIs, which I think is available already, see https://github.com/wildfly/wildfly/blob/master/pom.xml#L368.

Anyway, about blocking issues, you mean the only relatin is for the following specs:

https://projects.eclipse.org/projects/technology.microprofile/releases/4.0

to end up in WF 23 together, right?

File: microprofile/WFLY-14281_Integrate-MP-Fault-Tolerance-3.0.adoc (hunk 4 above)
Comment (fabiobrz) [reply microprofile/WFLY-14281_Integrate-MP-Fault-Tolerance-3.0.adoc:62]: Ok, thanks

File: microprofile/WFLY-14281_Integrate-MP-Fault-Tolerance-3.0.adoc (hunk 5 above)
Comment (fabiobrz) [reply microprofile/WFLY-14281_Integrate-MP-Fault-Tolerance-3.0.adoc:77]: Ok, thanks

**Comment by fabiobrz** [comment]
Thanks for addressing my comments @rhusar - I've dropped a couple more questions.

File: microprofile/WFLY-14281_Integrate-MP-Fault-Tolerance-3.0.adoc (hunk 3 above)
Comment (rhusar) [reply microprofile/WFLY-14281_Integrate-MP-Fault-Tolerance-3.0.adoc:49]: No (at least I haven't found anything that needs updating).

File: microprofile/WFLY-14281_Integrate-MP-Fault-Tolerance-3.0.adoc (hunk 2 above)
Comment (rhusar) [reply microprofile/WFLY-14281_Integrate-MP-Fault-Tolerance-3.0.adoc:32]: You need to look into the 'MP4' branch, so: https://github.com/wildfly/wildfly/blob/MP4/pom.xml#L369 which is already updated.

Yes, MP 4.0 needs to go in together as there are dependencies of specs on one another. There is no compatibility to take something from MP3 and use it with MP4 libs.

Comment (fabiobrz) [reply microprofile/WFLY-14281_Integrate-MP-Fault-Tolerance-3.0.adoc:32]: Thanks. I looked at the `master` in order to double check that Metrics 2.* had been used already there.

**Review by fabiobrz — CHANGES_REQUESTED** [review]
Hi @rhusar - thanks for answering previous questions. I've dropped one comment regarding a link format.

File: microprofile/WFLY-14281_Integrate-MP-Fault-Tolerance-3.0.adoc [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+== Overview
+
+MicroProfile 4.0 provides a major upgrade for https://github.com/eclipse/microprofile-fault-tolerance/releases/tag/3.0[MicroProfile Fault Tolerance 3.0] with changes described in its
+https://download.eclipse.org/microprofile/microprofile-fault-tolerance-3.0/microprofile-fault-tolerance-spec-3.0.html#release_notes_30[Release Notes].
+
+This component upgrade also covers the upgrade of SmallRye Fault Tolerance that implements MicroProfile Fault Tolerance 3.0.
+
+The spec itself made several backward incompatible changes, which can be found
+here[https://download.eclipse.org/microprofile/microprofile-fault-tolerance-3.0/microprofile-fault-tolerance-spec-3.0.html#_backward_incompatible_changes],
```
[the remark this answers is not in the document]
Comment (rhusar) [reply microprofile/WFLY-14281_Integrate-MP-Fault-Tolerance-3.0.adoc]: Fixed, thanks.

**Review by fabiobrz — APPROVED** [review]
Thanks @rhusar. I am approving the proposal.

**Pull request merged**